GVFS/GVFS.Common/GVFSEnlistmentStatistics.cs

```diff
@@ -0,0 +1,189 @@
+    {
+        private List<string> gitPaths;
+        private List<string> placeholderFolderPaths;
+        private List<string> placeholderFilePaths;
+        private List<string> modifiedFolderPaths;
+        private List<string> modifiedFilePaths;
+        private int trackedFilesCount;
+        private int placeholderCount;
+        private int modifiedPathsCount;
+        private Dictionary<string, int> trackedFilesDirectoryTracking;
+        private Dictionary<string, int> hydratedFilesDirectoryTracking;
+
+        public GVFSEnlistmentStatistics(List<string> gitPaths, List<string> placeholderFolderPaths, List<string> placeholderFilePaths, List<string> modifiedFolderPaths, List<string> modifiedFilePaths)
+        {
+            this.gitPaths = gitPaths;
+            this.placeholderFolderPaths = placeholderFolderPaths;
+            this.placeholderFilePaths = placeholderFilePaths;
+            this.modifiedFolderPaths = modifiedFolderPaths;
+            this.modifiedFilePaths = modifiedFilePaths;
+        }
+
+        public void CalculateStatistics(string parentDirectory)
+        {
+            this.trackedFilesCount = 0;
+            this.placeholderCount = 0;
+            this.modifiedPathsCount = 0;
+            this.trackedFilesDirectoryTracking = new Dictionary<string, int>();
+            this.hydratedFilesDirectoryTracking = new Dictionary<string, int>();
+
+            this.CategorizePaths(this.gitPaths, this.trackedFilesDirectoryTracking, ref this.trackedFilesCount, parentDirectory, isFile: true);
+            this.CategorizePaths(this.placeholderFolderPaths, this.hydratedFilesDirectoryTracking, ref this.placeholderCount, parentDirectory, isFile: false);
+            this.CategorizePaths(this.placeholderFilePaths, this.hydratedFilesDirectoryTracking, ref this.placeholderCount, parentDirectory, isFile: true);
+            this.CategorizePaths(this.modifiedFolderPaths, this.hydratedFilesDirectoryTracking, ref this.modifiedPathsCount, parentDirectory, isFile: false);
+            this.CategorizePaths(this.modifiedFilePaths, this.hydratedFilesDirectoryTracking, ref this.modifiedPathsCount, parentDirectory, isFile: true);
+        }
+
+        public List<string> GetDirectoriesSortedByHydration()
+        {
+            List<string> directoriesSortedByHydration = new List<string>();
+
+            foreach (KeyValuePair<string, int> keyValuePair in this.trackedFilesDirectoryTracking)
+            {
+                directoriesSortedByHydration.Add(keyValuePair.Key);
+            }
+
+            directoriesSortedByHydration.Sort(
+                delegate(string directoryOne, string directoryTwo)
+                {
+                    return this.GetHydrationOfDirectory(directoryOne).CompareTo(this.GetHydrationOfDirectory(directoryTwo)) * -1; // Multiply by -1 to reverse the sort
+                });
+
+            return directoriesSortedByHydration;
+        }
+
+        public double GetHydrationOfDirectory(string directory)

```

Comment:
Under this current system, yes. It could be expanded to include the other file tree information but in an attempt to not just create the entire file tree in memory, right now only the directories immediately under the one the utility was run on can be accessed